src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataReaderExtensions.cs

```diff
@@ -71,7 +72,45 @@ public static ODataItemBase ReadResourceOrResourceSet(this ODataReader reader)
                             "The resource which is ending should be on the top of the items stack.");
                         itemsStack.Pop();
                         break;
+                    case ODataReaderState.DeletedResourceStart:
+                        ODataDeletedResource deletedResource = (ODataDeletedResource)reader.Item;
+                        ODataResourceWrapper deletedResourceWrapper = null;
+                        if (deletedResource != null)
+                        {
+                            deletedResourceWrapper = new ODataResourceWrapper(deletedResource);
+                        }
 
+                        if (itemsStack.Count == 0)
+                        {
+                            Contract.Assert(deletedResource != null, "The top-level resource can never be null.");
+                            topLevelItem = deletedResourceWrapper;
+                        }
+                        else
+                        {
+                            ODataItemBase parentItem = itemsStack.Peek();
+                            ODataResourceSetWrapper parentResourceSet = parentItem as ODataResourceSetWrapper;
+                            if (parentResourceSet != null)
+                            {
+                                parentResourceSet.Resources.Add(deletedResourceWrapper);
+                            }
+                            else
+                            {
+                                ODataNestedResourceInfoWrapper parentNestedResource = (ODataNestedResourceInfoWrapper)parentItem;
+                                Contract.Assert(parentNestedResource.NestedResourceInfo.IsCollection == false, "Only singleton nested properties can contain resource as their child.");
+                                Contract.Assert(parentNestedResource.NestedItems.Count == 0, "Each nested property can contain only one resource as its direct child.");
+                                parentNestedResource.NestedItems.Add(deletedResourceWrapper);
+                            }
```

Comment:
Consider making this a common routine "AddResourceToParent", called from both here and replacing 51-63 in ResourceStart above.